library/network/src/modules/SuSEFirewall.rb

```diff
@@ -1249,9 +1249,75 @@ def GetProtectFromInternalZone
     # @example GetInterfacesInZone ("external") -> ["eth4", "eth5"]
     def GetInterfacesInZone(zone)
       return [] if !GetKnownFirewallZones().include?(zone)
-      deep_copy(@SETTINGS[zone][:interfaces])
+      # Only return the known ones
+
+      interfaces_in_zone = deep_copy(@SETTINGS[zone][:interfaces])
+
+      known_interfaces_now = GetListOfKnownInterfaces()
+
+      # filtering special strings
+      interfaces_in_zone = Builtins.filter(interfaces_in_zone) do |interface|
+        interface != "" && Builtins.contains(known_interfaces_now, interface)
+      end
+
+      deep_copy(interfaces_in_zone)
     end
 
+    # Function removes interface from defined zone.
+    #
+    # @param [String] interface
+    # @param [String] zone
+    # @example RemoveInterfaceFromZone ("modem0", "EXT")
+    def RemoveInterfaceFromZone(interface, zone)
+
+      return nil if !IsKnownZone(zone)
+
+      SetModified()
+
+      Builtins.y2milestone(
+        "Removing interface '%1' from '%2' zone.",
+        interface,
+        zone
+      )
+
+      @SETTINGS[zone][:interfaces].delete(interface)
+      @SETTINGS[zone][:modified] << :interfaces
+
+      nil
+    end
+
+    # Functions adds interface into defined zone.
+    # All appearances of interface in other zones are removed.
+    #
+    # @param [String] interface
+    # @param [String] zone
+    # @example AddInterfaceIntoZone ("eth5", "DMZ")
+    def AddInterfaceIntoZone(interface, zone)
+
+      return nil if !IsKnownZone(zone)
+
+      SetModified()
+
+      current_zone = GetZoneOfInterface(interface)
+
+      # removing all appearances of interface in zones, excepting current_zone==new_zone
+      while !current_zone.nil? && current_zone != zone
```

Comment:
That ` && current_zone != zone` is a bug, it will bail out early even if other zones include our interface but happen to come after *zone* in the listing.